ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,308 @@
+            self.instrument_table = table
+
+        self.shower_table = None
+        if self.load_simulated and SHOWER_TABLE in self.h5file:
+            self.shower_table = read_table(self.h5file, SHOWER_TABLE)
+
+    def close(self):
+        self.h5file.close()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _read_telescope_table(self, group, tel_id):
+        if self.structure == "by_id":
+            key = f"{group}/tel_{tel_id:03d}"
+            condition = None
+        else:
+            key = f"{group}/{self.subarray.tel[tel_id]!s}"
+            condition = f"tel_id == {tel_id}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=condition)
+        else:
+            table = Table()
+
+        return table
+
+    def read_subarray_events(self, table=None):
+
+        if self.shower_table:
+            if (table is not None) and len(table) > 0:
+                table = join(
+                    table,
+                    self.shower_table,
+                    keys=["obs_id", "event_id"],
+                    join_type="inner"
+                )
+            else:
+                table = self.shower_table
+
+        if self.load_trigger:
+            if (table is not None) and len(table) > 0:
+                table = join(
+                    table,
+                    read_table(self.h5file, TRIGGER_TABLE),
+                    keys=["obs_id", "event_id"],
+                    join_type="inner",
+                )
+            else:
+                table = read_table(self.h5file, TRIGGER_TABLE)
+
+        if self.load_dl2_geometry:
+
+            shower_geometry_group = self.h5file.root[GEOMETRY_GROUP]
+
+            for i, reconstructor in enumerate(shower_geometry_group._v_children):
+
+                geometry = read_table(self.h5file, f"{GEOMETRY_GROUP}/{reconstructor}")
+
+                # rename DL2 columns to explicit reconstructor
+                # TBD: we could skip this if only 1 reconstructor is present
+                # or simply find another way to deal with multiple reconstructions
+                for col in set(geometry.colnames) - {"obs_id", "event_id"}:
+                    geometry.rename_column(col, f"{reconstructor}_{col}")
+
+                if ((table is None) or (len(table) == 0)) and (i == 0):
+                    table = geometry
+                else:
+                    table = join(
+                        table, geometry, keys=["obs_id", "event_id"], join_type="left"
+                    )
+        return table
+
+    def read_telescope_events(self, tel_ids):
+
+        table = vstack([self.read_telescope_events_for_id(tel_id) for tel_id in tel_ids])
+
+        return table
+
+    def read_events(self, labels=None):
+        """Read telescope-based event information.
```

Comment:
Why not read_telescope_events? (the other is `read_subarray_events`).  this confused me, since i expected something only called events to load the shower events, not the telescope ones.

Reply:
also what happens here if I mix two telescope types and have `load_images=True`?  If it fails, this should be explained that it doesn't work for mixed cameras if the number of pixels is different.   Or perhaps this could be a private method and we only expose `read_events_by_tel_type`?

Reply:
we have now both `read_telescope_events`, which gives tables based on single telescope data and `read_subarray_events` which reads only subarray data

`read_events` is simply the general function used to call the first (if DL1 loadings have been enabled) and then the 2nd (if subarray loadings have been enabled

Reply:
yes it fails if we try to stack images of different sizes

do we have a nicer solution for this?

Reply:
There is no easy solution.  you could just make this private since it is a helper functions used by the higher-level functions. Otherwise, you need to add a check and raise a friendly exception explaining why it doesn't work.

Reply:
well it doesn't matter if you make it a private method no? if someone wants a single-table with true images from different cameras it will always fail

I think we should raise an error (and also specify it in the docs with a disclaimer)

did I misunderstood what you meant?

Reply:
well, if it's private it is not part of the public API, so users should not use it.  Instead they would use functions that call it..  But keeping it and adding an error handler is also fine.